guardrails-ai/guardrails-api
# added open search to docker compose

Notes/Considerations:

- **Testing**
  - `docker-compose -f 'docker-compose-dev.yml' up -d --build`
  - Add api key where INSERT and getting-started.rail to directory
  - /ingest
  - In OpenSearch, create index patterns for events* and metrics* to see logs and metrics respectively. This can be done under Stack Management. These will then be viewable in the Discover tab. Traces will populate in the Traces section under Observability in the side panel
- Metrics do not associate with a trace id in OpenSearch even if nested under a span. May need to look into manually configuring the trace and span ids
- OpenAPI spec (specifically for apis around logs and traces) may be outdated after seeing what traces/logs look like in OpenSearch
- **Not a priority**: Both Metrics and Logs can be visualized more deeply using Prometheus and Fluent Bit respectively. These will have to be added to the docker compose, pipeline.yml and otel-collector-config.yml
- **Not a priority**: Need to consider adding ssl/tls security to Otel Collector, DataPrepper, and OpenSearch configs
  - Otel Collector: https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/configtls/README.md
  - Data Prepper: https://opensearch.org/docs/latest/data-prepper/managing-data-prepper/configuring-data-prepper/
  - OpenSearch: https://opensearch.org/docs/latest/install-and-configure/install-dashboards/tls/ 

Task Timeline

1. Add [otel-instrumentors](https://github.com/tinystacks/otel-instrumentors) as a dependency to both guardrails-poc and guardrails sdk integration. Currently, we are using modules inside of Guardrails POC to gather metrics but this will not work when integrating with guardrails sdk. [This pr](https://github.com/ShreyaR/guardrails/pull/215) shows where we need to add traces and logging in guardrails sdk for validate api calls
2. Move OpenSearch client out of app.py and into the telemetry api (you can find an example query in app.py)
3. Add vector projections calculated through t-SNE and other metrics to OpenSearch
4. Add telemetry api requests for metrics that return plottable data points
